### PR TITLE
Bug fixes. 

### PR DESCRIPTION
**Pull request by DanielKrawisz**
* Fixed bug in objectmanager that caused lots of peers to be disconnected at once due to expired object requests. 
* Fixed bug in database that sometimes returned nil hashes in RemoveExpiredObjects.
* Moved the test objects to their own files. 
* Some work simplifying the tests slightly. 
* An earlier bug fix created a condition in peer tests which sometimes caused all goroutines to lock up. I fixed the tests. 
* Rarely the rpc server test would fail if two connections were both started at the same time before the connection was actually read by either. I fixed the test so that the test does not fail, but it is still possible for this rare error case to happen if two connections are made at near enough the same time.

### CODE REVIEW

File: objectmanager.go [hunk 1]
```diff
@@ -424,6 +435,13 @@ func (om *ObjectManager) clearRequests(d time.Duration) {
 				p.timestamp.Add(d*time.Duration(peerQueueSize)).Before(now) ||
 				p.peer.lastReceipt.Add(d).Before(now) {
 
+				// If it took more than ten minutes for the object to be requested,
+				// then it may have just expired, do just drop the request.
+				if p.knownSince.Add(10 * time.Minute).Before(p.timestamp) {
+					delete(om.requested, hash)
+					continue
+				}
```
Comment (ishbir) [inline objectmanager.go]: Comment is a little misleading. In the current form, it implies that the process of requesting the object took more than 10 mins. Should be something like: "If more than ten minutes have passed since the object was requested, then it may have just expired. So just drop the request." Also, could you make the 10 in peerQueueSize < 10, and the object drop duration unexported constants?

Comment (DanielKrawisz) [inline objectmanager.go]: Do you really think it's necessary to turn these into config options? These numbers are only relevant during a fairly unusual circumstance, ie, when bmd is first turned on with an empty database. I don't think that someone would actually want to configure them and all the extra directions in the --help menu will just confuse people.

Comment (ishbir) [inline objectmanager.go]: Oh no.. just unexported constants at the top of the file. Makes it easier to change things (for the dev, and not user) when needed.

Comment (DanielKrawisz) [inline objectmanager.go]: Oh I misunderstood. Yes I'll get to that right now.

File: objectmanager.go [hunk 2]
```diff
@@ -33,6 +33,16 @@ const (
...
+
+	// unsentObjectPenaltyTimeout is the duration of time it takes before an
+	// object the request for which has timed out to be considered more likely
+	// to have expired than maliciously withheld by the peer it was requested
+	// from.
+	unsentObjectPenaltyTimeout = 10 * time.Minute
+
+	// maxQueueForRequestTimeoutPenalty is the minimum number of requests a peer
+	// must have before object request timeouts are treated more leniantly.
+	maxQueueForRequestTimeoutPenalty = 10
```
Comment (ishbir) [inline objectmanager.go]: Could you rephrase comment for unsentObjectPenaltyTimeout? It's confusing to read. Also, s/leniantly/leniently/.

File: peer/connection.go [hunk 3]
```diff
@@ -79,6 +79,8 @@ func (pc *connection) WriteMessage(msg wire.Message) error {
 		return err
 	}
 
+	pc.idleTimer.Reset(pc.idleTimeout)
+
```
Comment (ishbir) [inline peer/connection.go]: Why'd you reset this again? It's already been done on line 71.

File: peer/peer.go [hunk 4]
```diff
@@ -169,7 +169,7 @@ out:
 		case *wire.MsgGetData:
 			err = p.logic.HandleGetDataMsg(msg)
 
-		case *wire.MsgGetPubKey, *wire.MsgPubKey, *wire.MsgMsg, *wire.MsgBroadcast, *wire.MsgUnknownObject:
+		case *wire.MsgGetPubKey, *wire.MsgPubKey, *wire.MsgMsg, *wire.MsgBroadcast, *wire.MsgUnknownObject, *wire.MsgObject:
```
Comment (ishbir) [inline peer/peer.go]: Why did you add this here? There's no way that the decoded message could be of *wire.MsgObject. It's converted into one of the other five types. This ensures that there's proper validation of respective object payloads.

File: peer_test.go [hunk 5]
```diff
@@ -1123,6 +385,8 @@ func TestInboundPeerHandshake(t *testing.T) {
 
 		serv.WaitForShutdown()
 	}
+
+	resetCfg(cfg)()
```
Comment (ishbir) [inline peer_test.go]: Why'd you add this?

File: peer_test.go [hunk 6]
```diff
@@ -1178,22 +442,26 @@ func TestProcessAddr(t *testing.T) {
...
 		{
-			&PeerAction{nil, nil, true, false},
+			&PeerAction{InteractionComplete: true},
 			0,
 		},
 	}
 
 	for testCase, addrTest := range AddrTests {
-		defer resetCfg(cfg)()
-
```
Comment (ishbir) [inline peer_test.go]: Why did you remove this?

File: peer_test.go [hunk 7]
```diff
@@ -1229,65 +497,8 @@ func TestProcessAddr(t *testing.T) {
...
-}
 
-func NewMockSendQueue(mockConn *MockConnection) *MockSendQueue {
-	return &MockSendQueue{
-		conn:         mockConn,
-		msgQueue:     make(chan wire.Message, 1),
-		requestQueue: make(chan []*wire.InvVect, 1),
-		quit:         make(chan struct{}),
-	}
+	resetCfg(cfg)()
```
Comment (ishbir) [inline peer_test.go]: Why'd you add this?

File: peer_test.go [hunk 8]
```diff
@@ -1377,8 +588,6 @@ func TestProcessInvAndObjectExchange(t *testing.T) {
 	remoteAddr := &net.TCPAddr{IP: net.ParseIP("192.168.0.1"), Port: 8333}
 
 	for testCase, test := range tests {
-		defer resetCfg(cfg)()
-
```
Comment (ishbir) [inline peer_test.go]: Same.

File: peer_test.go [hunk 9]
```diff
@@ -1415,5 +624,7 @@ func TestProcessInvAndObjectExchange(t *testing.T) {
 				}
 			}
 		}
+
+		resetCfg(cfg)()
```
Comment (ishbir) [inline peer_test.go]: Same.

File: rpcserver_test.go [hunk 10]
```diff
@@ -252,8 +253,8 @@ func TestRPCConnection(t *testing.T) {
 	// Create a server.
 	listeners := []string{net.JoinHostPort("", "8445")}
 	serv, err = newServer(listeners, getMemDb([]*wire.MsgObject{}),
-		MockListen([]*MockListener{
-			NewMockListener(remoteAddr, make(chan peer.Connection), make(chan struct{}, 1))}))
+		bmdtest.MockListen([]*bmdtest.MockListener{
+			bmdtest.NewMockListener(remoteAddr, make(chan peer.Connection), make(chan struct{}, 1))}))
```
Comment (ishbir) [inline rpcserver_test.go]: Could you remove these from bmdtest as well?

File: testobjects_test.go [hunk 11]
```diff
@@ -0,0 +1,812 @@
...
+
+// RandomShaHash returns a ShaHash with a random string of bytes in it.
+func RandomShaHash() *wire.ShaHash {
+	b := make([]byte, 32)
+	for i := 0; i < 32; i++ {
+		b[i] = byte(rand.Intn(256))
+	}
+	hash, _ := wire.NewShaHash(b)
+	return hash
+}
```
Comment (ishbir) [inline testobjects_test.go]: It doesn't need to be an exported function. It would have been fine if you left it as randomShaHash as well. :)

File: peer/testobjects_test.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+// RandomShaHash returns a ShaHash with a random string of bytes in it.
+func RandomShaHash() *wire.ShaHash {
+	b := make([]byte, 32)
+	for i := 0; i < 32; i++ {
+		b[i] = byte(rand.Intn(256))
+	}
+	hash, _ := wire.NewShaHash(b)
+	return hash
+}
```
Comment (ishbir) [inline peer/testobjects_test.go]: It would have been fine as randomShaHash. There was no need for a change. :)

File: peer/connection.go (hunk 3 above)
Comment (DanielKrawisz) [inline peer/connection.go]: I think this is a mistake caused by a git rebase that happened at some point.

File: peer/peer.go (hunk 4 above)
Comment (DanielKrawisz) [inline peer/peer.go]: You're right. It shouldn't be there. I think that I added an object type to a test case. I'll see if I can make the test work correctly and have the object type be ignored.

File: peer_test.go (hunk 5 above)
Comment (DanielKrawisz) [inline peer_test.go]: Also something to do with git.

File: objectmanager.go [hunk 13]
```diff
@@ -62,10 +79,18 @@ type donePeerMsg struct {
...
 	timestamp time.Time
+	// The time that the inv was first heard about.
+	knownSince time.Time
 }
 
+// Whether to disconnect a peer for not handling the request. It can happen
+// that a hash is stored in the queue for so long that there is a good chance
+// the corresponding object has expired by the time it was requested. In this
+// case, we don't want to penalize peers
+
```
Comment (ishbir) [inline objectmanager.go]: What's this comment for?

File: objectmanager.go [hunk 14]
```diff
@@ -420,10 +448,17 @@ func (om *ObjectManager) clearRequests(d time.Duration) {
 			// If the peer has a long queue of requested objects, then don't
 			// disconnect it unless it has not received SOME object recently enough,
 			// or if the request is expired by a much longer time.
-			if peerQueueSize < 10 ||
+			if peerQueueSize < maxQueueForRequestTimeoutPenalty ||
 				p.timestamp.Add(d*time.Duration(peerQueueSize)).Before(now) ||
 				p.peer.lastReceipt.Add(d).Before(now) {
 
+				// If it took more than ten minutes for the object to be requested,
+				// then it may have just expired, do just drop the request.
```
Comment (ishbir) [inline objectmanager.go]: Could you change comment to "If more than ten minutes have passed since the object was requested, it may have just expired. Drop the request."? It's a little confusing in its current form, especially "If it took more than ten minutes for the object to be requested".

File: rpcserver_test.go [hunk 15]
```diff
@@ -301,6 +301,8 @@ func TestRPCConnection(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	time.Sleep(20 * time.Millisecond)
```
Comment (ishbir) [inline rpcserver_test.go]: Could you add a brief comment stating why you added this?

File: objectmanager.go (hunk 13 above)
Comment (DanielKrawisz) [inline objectmanager.go]: That was a draft of a comment that I accidentally left in when I was just thinking about what to say.
